components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedIdPData.java

```diff
@@ -46,11 +64,107 @@ public void setUser(AuthenticatedUser user) {
+     * @return
+     */
+    @Deprecated
     public AuthenticatorConfig getAuthenticator() {
-        return authenticator;
+
+        // If the serialized authenticated data has been stored with the old class definition,
+        // 'authenticator' field may be available. So that variable takes the priority over 'authenticators' list.
+        if (authenticator != null) {
+            if (log.isDebugEnabled()) {
+                log.debug("Serialized and stored AuthenticatedIdPData object was initially serialized using the " +
+                        "old class definition. Handling it in a backward compatible manner");
+            }
+            return authenticator;
+        } else if (CollectionUtils.isNotEmpty(authenticators)) {
+            // NOTE : In order to make introducing 'authenticators' field, backward compatible
+            // setAuthenticator() and getAuthenticator() methods should remain working as before.
+            return authenticators.get(authenticators.size() - 1);
+        } else {
+            return null;
+        }
+
     }
 
+    /**
+     * @deprecated use {@link #addAuthenticator(AuthenticatorConfig)} instead.
+     * @param authenticator
+     */
+    @Deprecated
     public void setAuthenticator(AuthenticatorConfig authenticator) {
-        this.authenticator = authenticator;
+        // NOTE : In order to make introducing 'authenticators' backward compatible.
+        // setAuthenticator() and getAuthenticator() methods should remain working as before.
+        addAuthenticator(authenticator);
+    }
+
+    /**
+     * Adds the given authenticator as an authenticator which the user has authenticated with.
+     *
+     * @param authenticator Authenticator config to be added.
+     */
+    public void addAuthenticator(AuthenticatorConfig authenticator) {
+        // If the serialized authenticated data has been stored with the old class definition,
+        // authenticators' field will be null, when loading this object from the data store.
+        if (this.authenticators == null) {
+            this.authenticators = new ArrayList<AuthenticatorConfig>();
+        }
+        this.authenticators.add(authenticator);
+    }
+
+    /**
+     * Returns the authenticators which were used to authenticated the user.
+     *
+     * @return the authenticators which were used to authenticated the user
+     */
+    public List<AuthenticatorConfig> getAuthenticators() {
+
+        // If the serialized authenticated data has been stored with the old class definition,
+        // 'authenticator' field might be not null and 'authenticators' field will be null.
+        // So merge 'authenticator' object with the 'authenticators' list accordingly before returning.
+
+        List<AuthenticatorConfig> authenticatorsToBeReturned = new ArrayList<AuthenticatorConfig>();
+
+        if (this.authenticators != null) {
+            authenticatorsToBeReturned = (List<AuthenticatorConfig>) ((ArrayList) this.authenticators).clone();
```

Comment:
Reassigning already created "authenticatorsToBeReturned", which means we create un-necessary list at line 128.
Better to use " new ArrayList<AuthenticatorConfig>(this.authenticators);" instead of clone.

Reply:
How about this? 
---
if (this.authenticators != null) {
return new ArrayList(this.authenticators);
}else {
return Arrays.asList(this.authenticator));
}